spec/prog/kubernetes/provision_kubernetes_node_spec.rb

```diff
@@ -0,0 +1,212 @@
+require_relative "../../model/spec_helper"
+
+RSpec.describe Prog::Kubernetes::ProvisionKubernetesNode do
+  subject(:prog) { described_class.new(Strand.new) }
+
+  let(:kubernetes_cluster) {
+    project = Project.create_with_id(name: "default").tap { _1.associate_with_project(_1) }
+    subnet = PrivateSubnet.create_with_id(net6: "0::0/16", net4: "127.0.0.0/8", name: "x", location: "x", project_id: project.id)
+    kc = KubernetesCluster.create_with_id(
+      name: "k8scluster",
+      kubernetes_version: "v1.32",
+      cp_node_count: 3,
+      private_subnet_id: subnet.id,
+      location: "hetzner-fsn1",
+      project_id: project.id
+    )
+    kc.associate_with_project(project)
+    subnet.associate_with_project(project)
+
+    lb = LoadBalancer.create_with_id(private_subnet_id: subnet.id, name: "somelb", src_port: 123, dst_port: 456, health_check_endpoint: "/foo")
+    kc.add_vm(create_vm)
+    kc.add_vm(create_vm)
+
+    kc.update(api_server_lb_id: lb.id)
+    kc
+  }
+
+  before do
+    allow(prog).to receive(:kubernetes_cluster).and_return(kubernetes_cluster)
+    allow(prog).to receive(:frame).and_return({"vm_id" => create_vm.id})
+  end
+
+  describe "#write_hosts_file_if_needed" do
+    it "exits early if the environment is not dev" do
+      expect(prog.vm).not_to receive(:sshable)
+      expect(Config).to receive(:development?).and_return(false)
+      prog.write_hosts_file_if_needed
+    end
+
+    it "exits early if /etc/hosts file contains an entry about the cluster endpoint already" do
+      sshable = instance_double(Sshable)
+      allow(prog.vm).to receive(:sshable).and_return(sshable)
```

Comment:
why allow here? I only use allow in `before` blocks if it really helps with some difficult combination of methods to stub; I use `expect` so that when it goes obsolete, it complains and I can delete it. `allow` inevitably converges on plaque that accumulates.